engine/src/main/java/org/hibernate/validator/internal/engine/groups/ExpandedGroup.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.engine.groups;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+
+/**
+ * Represents a validation group and all the groups it extends ("group inheritance").
+ *
+ * @author Gunnar Morling
+ */
+public class ExpandedGroup implements Iterable<Group> {
```

Comment:
> I tried that first actually

Ahh. Good we thought alike.

> but it didn't work out eventually.

Interesting. 

> I still think it'd work in some way, but it'd need more investigation and changes

I see. Personally I think I would have investigate a bit more, but if you think this is the best approach for now, let's go with it.

> That said, I am all for revisting the entire implementation at some point

+1

>  I got some ideas, but it'll be a larger piece of work I am afraid.

Maybe we can talk about it over a chat?